File: workspace-service/src/main/java/com/app/workspace_service/service/WorkspaceService.java
package com.app.workspace_service.service;

import java.util.List;
import java.util.UUID;

import com.app.workspace_service.dto.BookingRequest;
import com.app.workspace_service.dto.WorkspaceRequest;
import com.app.workspace_service.entity.Booking;
import com.app.workspace_service.entity.Slot;
import com.app.workspace_service.entity.Workspace;

public interface WorkspaceService {

   Workspace createWorkspace(WorkspaceRequest request, String userId);

   void bookSlot(BookingRequest request, String userId);

   List<Workspace> getWorkspaces(String location);

   List<Slot> getSlots(UUID workspaceId);

   List<Booking> getParticipants(UUID slotId);

   void cancelBooking(UUID bookingId, String userId);

   void approveBooking(UUID bookingId, String userId);

   void rejectBooking(UUID bookingId, String userId);
}

File: workspace-frontend/src/api/workspaceService.js
import workspaceClient from "./workspaceClient"

/**
 * Fetch all workspaces (with optional location filter)
 */
export const getAllWorkspaces = async (location) => {
  const res = await workspaceClient.get("/workspace", {
    params: { location },
  })
  return res.data
}

/**
 * Create new workspace
 */
export const createWorkspace = async (payload) => {
  const res = await workspaceClient.post("/workspace", payload)
  return res.data
}

/**
 * ✅ Fetch slots for a workspace
 */
export const getSlots = async (workspaceId) => {
  console.log("Calling slots API with:", workspaceId)

  const res = await workspaceClient.get(
    `/workspace/${workspaceId}/slots`
  )

  return res.data
}

/**
 * ✅ Book workspace (creates OR joins slot)
 */
export const bookWorkspace = async (payload) => {
  const res = await workspaceClient.post("/workspace/book", payload)
  return res.data
}

export const getWorkspaceById = async (id) => {
  const response = await workspaceClient.get(`/workspace/${id}`)
  return response.data
}

export const getParticipants = async (slotId) => {
  const res = await workspaceClient.get(
    `/workspace/slot/${slotId}/participants`
  )
  return res.data
}

export const approveBooking = async (bookingId) => {
  console.log("Calling approve API with:", bookingId);
    const res = await workspaceClient.post(
        `/workspace/booking/${bookingId}/approve`
    )
    return res.data
}

export const rejectBooking = async (bookingId) => {
    const res = await workspaceClient.post(
        `/workspace/booking/${bookingId}/reject`
    )
    return res.data
}